[RFC] 11-pkglint-elf-in-usrshare: allow explicit setting of exceptions

Hoshpak opened the pull request:
This will allow to explicitely ignore files in the hook where it otherwise would create false positives like in the case of qemu.

@void-linux/pkg-committers

ahesford commented on common/hooks/post-install/11-pkglint-elf-in-usrshare.sh:
```diff
@@ -18,7 +18,11 @@ hook() {
             # Note application/x-executable is missing which is present in most Electron apps
             application/x-sharedlib*|application/x-pie-executable*)
-                matches+=" ${file#$PKGDESTDIR}" ;;
+                if ! [[ ${ignore_in_usr_share} =~ $(basename ${file#$PKGDESTDIR}) ]]
```
A test like
```
[[ ${ignore_in_usr_share} != *$(basename "$I")* ]]
```
to the match operator would be consistent with with match tests in the build-helpers.

Regardless, the argument to `basename` should be quoted in case the file has spaces in the name.

Also, why bother stripping out `PKGDESTDIR` from `file` if we are feeding it to basename anyway?

pullmoll commented:
Clean solution LGTM

Hoshpak commented:
Am 16.07.20 um 20:30 schrieb Andrew J. Hesford:
> *@ahesford* requested changes on this pull request.
> 
> I think we will always have difficulty with files that might contain spaces in 
> the name, because we will not be able to distinguish a file named |File with 
> spaces| in |$ignore_in_usr_share| and three separate files named |File|, |with| 
> and |spaces|. Using an array instead of space-separated strings avoids this, but 
> gets clumsy in the template (if we allow arrays in |ignore_in_usr_share|, why 
> not for any other variables?).

I think the general consensus up to know was to keep arrays out of
xbps-src. It will make the code less portable but we use a lot of
bashims already and the general assumption is that it will be run with
bash. Personally I just dislike how arrays are handled in bash and
usually tend to rather rewrite a script in python than having to deal
with them.

> It might also be nice to support directories in |ignore_in_usr_share| so we can 
> do things like |usr/share/qemu/firmware| instead of listing each file, but that 
> involves a second test and a bit of thought.

In the case of qemu there are just two files which make the hook fail. I
don't know if there are packages with a lot of files which would require
exceptions though.

> |[[ ${ignore_in_usr_share} != *$(basename "$I")* ]] |
> 
> to the match operator would be consistent with with match tests in the 
> build-helpers.
> 
> Regardless, the argument to |basename| should be quoted in case the file has 
> spaces in the name.

Thanks, I changed the check as you suggested and verified that it now
works for files with spaces.

> Also, why bother stripping out |PKGDESTDIR| from |file| if we are feeding it to 
> basename anyway?

Just an oversight on my part. I removed it.

Johnnynator commented on common/hooks/post-install/11-pkglint-elf-in-usrshare.sh:
```diff
@@ -18,7 +18,11 @@ hook() {
             # Note application/x-executable is missing which is present in most Electron apps
             application/x-sharedlib*|application/x-pie-executable*)
-                matches+=" ${file#$PKGDESTDIR}" ;;
+                if [[ ${ignore_in_usr_share} != *$(basename "${file}")* ]]
```
Can't you just compare it to `${file##/*}`, is faster but should give the same output as basename. (find does afaik never give you a path with a trailing `/`)

Piraty commented:
> In the case of qemu there are just two files which make the hook fail. I don't know if there are packages with a lot of files which would require exceptions though.

yes there are, i remember some microctrontroller firmware development package had to work around the hook really hard

Hoshpak commented:
> +                if [[ ${ignore_in_usr_share} != *$(basename "${file}")* ]]
> 
> Can't you just compare it to |${file##/*}|, is faster but should give the same 
> output as basename. (find does afaik never give you a path with a trailing |/|)

That works as well. I changed it.

Hoshpak commented:
>     In the case of qemu there are just two files which make the hook fail. I
>     don't know if there are packages with a lot of files which would require
>     exceptions though.
> 
> yes there are, i remember some microctrontroller firmware development package 
> had to work around the hook really hard

Then it is definitely beneficial to be able to exclude directories. I
split it into two variable `ignore_usrshare_files` and
`ignore_usrshare_dirs` in the lates iteration so we can either ignore
individual files or the directory containing the files.

ahesford commented:
Now that I'm thinking about this again, I wonder if looking at the basename is the right choice. Maybe it should just compare `${file#$PKGDESTDIR}` to the contents of `$ignore_usrshare_files` so we can differentiate files like
```
/usr/share/elf-allowed/myfile
/usr/share/elf-bad/myfile
```
The same should probably hold with `$ignore_usrshare_dirs`: require the full path (wrt `$PKGDESTDIR`) be specified instead of just the basename.

Chocimier commented:
Hook operates on usr/share so paths relative to usr/share feels natural. Then we can go crazy and not pass files in ignored directories to file(1). https://github.com/void-linux/void-packages/commit/da030cf5cffe3b845d24cbb9f29f5d12777a9eec

Hoshpak commented:
> Hook operates on usr/share so paths relative to usr/share feels natural. Then we can go crazy and not pass files in ignored directories to file(1). https://github.com/void-linux/void-packages/commit/da030cf5cffe3b845d24cbb9f29f5d12777a9eec
> 

Looks good, please got for it.

Chocimier commented:
There was idea to forbid elfs anywhere for noarch packages. In that case, full path would fit better.

> Looks good, please got for it.

When showing changes, I expected you take them into PR if liked them. Do you want me to overtake this PR instead?

Hoshpak commented:
>     Looks good, please got for it.
> 
> When showing changes, I expected you take them into PR if liked them. Do you 
> want me to overtake this PR instead?

Yes, please do so.

Chocimier commented:
Changed to absolute paths, added checking noarch packages, stepping over ignored directories, disallowed application/x-executable.

Chocimier commented:
Reverted check of noarch packages, described in manual. Will merge if no further comments.
